### ISSUE TITLE
allow_attributes_without_reason does nothing in Rust 1.81

### ISSUE THREAD

**Issue by danwilliams** [rust-lang/rust-clippy#13348]
### Summary

I have been using `clippy::allow_attributes_without_reason` for some time, with the `lint_reasons` feature enabled under nightly. Recently, [the `expect` issue was merged](https://github.com/rust-lang/rust/pull/120924#diff-77845d8a0dd9e03ec2b1933b108c109cf3b966709f09591042933ae4d29a988c), and I was excited to see that this would all be released as stable in Rust 1.81!

However, in trying this in Rust 1.81 (released today), I am unable to get it to work. This seems strange as there should be no feature gate now (obviously features are on nightly not stable), and nightly confirms the feature is stable. So why does it not trigger under the conditions it did before?

Adding the actual reasons is accepted, so it's just the lack of any error or warning from Clippy when they are absent which is puzzling! Perhaps I am doing something wrong?

Note, I have tried the lint via `Cargo.toml`, source code, and `-D` on the command line - nothing causes it to trigger. Example below is source code as that's simplest to illustrate.

### Reproducer

I tried this code:

```rust
#![forbid(clippy::allow_attributes_without_reason)]
#![deny(clippy::print_stdout)]
#![warn(clippy::allow_attributes)]

#[allow(clippy::print_stdout)]
fn main() {
    println!("Hello, world!");
}
```

I expected to see this happen: Clippy to tell me off for allowing without specifying a reason.

Instead, this happened: Clippy was happy, no errors. Note that I was also anticipating something about the use of allow instead of expect, but that didn't trigger either... but I've not previously used that feature.

### Version

```text
rustc 1.81.0 (eeb90cda1 2024-09-04)
binary: rustc
commit-hash: eeb90cda1969383f56a2637cbd3037bdf598841c
commit-date: 2024-09-04
host: aarch64-apple-darwin
release: 1.81.0
LLVM version: 18.1.7
```

**Comment by joshtriplett** [comment]
It seems like it would help to configure one of two potential behaviors:

1) Suppress lints that can only be fixed by upgrading to a new MSRV.
2) Emit such lints, with a note that fixing them will raise MSRV to 1.xy.

The former is useful for a project that wants to keep its MSRV at a given level. The latter is useful for a project that wants to *document* its MSRV but update it whenever needed.

**Comment by Jarcho** [comment]
Marking as nominated for @joshtriplett's suggestion. At RustConf right now and my laptop's keyboard is acting up so I can't write up a separate issue.

**Comment by Jarcho** [comment]
Closing in favor of #13416.
